src/twisted/internet/abstract.py

```diff
@@ -242,13 +242,17 @@ def doWrite(self):
 
         @see: L{twisted.internet.interfaces.IWriteDescriptor.doWrite}.
         """
-        if len(self.dataBuffer) - self.offset < self.SEND_LIMIT:
+        remaining = len(self.dataBuffer) - self.offset
+        if remaining < self.SEND_LIMIT:
             # If there is currently less than SEND_LIMIT bytes left to send
             # in the string, extend it with the array data.
-            self.dataBuffer = _concatenate(
-                self.dataBuffer, self.offset, self._tempDataBuffer
-            )
-            self.offset = 0
+            if remaining:
+                self.dataBuffer = _concatenate(
+                    self.dataBuffer, self.offset, self._tempDataBuffer
+                )
+                self.offset = 0
+            else:
+                self.dataBuffer = b"".join(self._tempDataBuffer)
```

Comment:
Why not update the offset here ?

```suggestion
            else:
                # All data from the currnet buffer was already sent.
                # Start with a new buffer. 
                self.dataBuffer = b"".join(self._tempDataBuffer)
```

Reply:
Eek. That's a bug! And a bug not caught by the test suite :cry:

Reply:
That's unfortunate ... but I think that is ok that it was observed here.

I hope that you can write a new tests that will fail ... and then get the fix implemented.

I think that this PR should not be merged, just yet.

needs-changes

Reply:
Yiiiikes, thank you for your attention to detail @adiroiban , this is a scary one reminiscent of some of the bugs we had in iocpreactor.